clients/drcachesim/simulator/cache_fifo.cpp

```diff
@@ -65,8 +65,16 @@ cache_fifo_t::init(int associativity, int block_size, int total_size,
 void
 cache_fifo_t::access_update(int block_idx, int way)
 {
-    // Since the FIFO replacement policy is independent of cache hit,
-    // we do not need to do anything here.
+    // When the current access is a hit and the currently accessed 'way' has a
+    // counter of 1, meaning it was the 'first accessed way' before this access,
+    // set the next 'way' as victim. This is because the currently accessed
+    // 'way' is no longer the 'first accessed way' and its counter is then
+    // cleared and the next 'way' is set as a victim.
```

Comment:
This is a breaking change that makes this no longer a FIFO replacement policy!  FIFO is supposed to be unaffected by frequency or recency of the last access: it is only about when the line was brought into the cache.

Reply:
I see. This was my understanding using the test case in `unit_test_cache_fifo_four_way()`
```
 // These accesses fill out all the 4 ways.
 cache_fifo_test.access_and_check_fifo(ADDRESS_A, 1); // A x X X
 cache_fifo_test.access_and_check_fifo(ADDRESS_B, 2); // A B x X
 cache_fifo_test.access_and_check_fifo(ADDRESS_C, 3); // A B C x
 cache_fifo_test.access_and_check_fifo(ADDRESS_D, 0); // a B C D

// The following three accesses are a hit (the last two being repeated accesses)
cache_fifo_test.access_and_check_fifo(ADDRESS_A, 1); 
cache_fifo_test.access_and_check_fifo(ADDRESS_A, 1); 
cache_fifo_test.access_and_check_fifo(ADDRESS_A, 1); 
```
After this hit shouldn't the next way to be replaced (next victim) be way 1?

Below is the accesses and their expected way to be replaced after each access with and without that change in `access_update()`

| Accessed Address |  Without change to `access_update()`              | With change to `access_update()` |
|--------------------------| --------------------------------------------------------------|----------------------------------------------
ADDRESS_A          |     1  |      1
ADDRESS_B          |    2   |     2
ADDRESS_C          |    3   |     3
ADDRESS_D          |    0   |     0
ADDRESS_A          |    0   |     1  
ADDRESS_A          |    0   |     1
ADDRESS_A          |    0   |     1
ADDRESS_E          |    0   |     2
ADDRESS_A          |    0   |     2

Reply:
> After this hit shouldn't the next way to be replaced (next victim) be way 1?

No, not for FIFO replacement.  Please see any description of FIFO replacement, e.g. wikipedia:
https://en.wikipedia.org/wiki/Cache_replacement_policies#First_in_first_out_(FIFO)
"The cache evicts the blocks in the order they were added, without any regard to how often or how many times they were accessed before."

Reply:
Gotcha. Removed the change I made in `access_update()` and updated the test cases and expected ways to be replaced after each access.